2.next Support Stacking Flash Messages

xhs345 opened the pull request:
In issue #7830 it was requested to backport the feature.

Commit:
Add support for stacking Flash messages

See Issue #7830

Commit:
Fix tests

markstory commented:
Looks good, I think changing the internal flash message format could be surprising in a bugfix release. This might have to wait until 2.10.0

xhs345 commented:
It's the nice kind of surprise, but I agree :)

markstory commented:
I've retargeted this pull request to 2.next, but you may have to rebase to clean up the history.

xhs345 commented:
Branch has been rebased

markstory commented on lib/Cake/Controller/Component/FlashComponent.php:
```diff
@@ -82,12 +82,18 @@ public function set($message, $options = array()) {
-		CakeSession::write('Message.' . $options['key'], array(
+		$messages = CakeSession::read('Message.' . $options['key']);
+
+		$newMessage = array(
```
In 3.x we support a `clear` option to allow for the flash message stack to be reset/cleared when adding a new message. Could we also have that option in 2.x? Without it maintaining existing behavior is not possible for those who want/need it.

markstory replied:
We also have `clear` as a class level configuration option which makes it easy to preserve existing behavior through a configuration change.

Commit:
Add clear option to Flash Message

To give user the option to disable Stacking of messages and being consistent with 3.x

xhs345 commented:
Yea, sounds like a good idea and was easy to add

ravage84 commented on lib/Cake/Controller/Component/FlashComponent.php:
```diff
@@ -38,6 +38,7 @@ class FlashComponent extends Component {
 		'key' => 'flash',
 		'element' => 'default',
 		'params' => array(),
+		'clear' => false
```
Haven't we agreed on a traling slash? :confused: 
Anyway, I would do so.

ravage84 replied:
Doh! Trailing comma, sorry :smirk_cat:

xhs345 replied:
I don't know what you agreed on, but it looks like a trailing comma is not being used in similar instances the other components currently

ravage84 commented on lib/Cake/View/Helper/FlashHelper.php:
```diff
@@ -82,10 +82,15 @@ public function render($key = 'flash', $options = array()) {
+		foreach ($flash as $message) {
+			$message['key'] = $key;
+			$message = $options + $message;
+			$out .= $this->_View->element($message['element'], $message);
```
Nitpicking: Could we add a new line after/before each new element? Would produce nice HTML.

xhs345 replied:
Right now the formatting is done in the view element and the HTML is looking quite nice by default imo 
![screenshot 2016-11-01 08 57 40](https://cloud.githubusercontent.com/assets/9390869/19896376/59461a20-a011-11e6-96a5-3da2d73d4461.png)

xhs345 replied:
Ok, guess I don't know which element and HTML you're referring to

ravage84 reviewed (approved):
Looks good, just some nitpicking...
